util/cpuprofile/pprof_api.go

```diff
@@ -0,0 +1,204 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cpuprofile
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/google/pprof/profile"
+	goutil "github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// ProfileHTTPHandler is same as pprof.Profile.
+// The difference is ProfileHTTPHandler uses cpuprofile.GetCPUProfile to fetch profile data.
+func ProfileHTTPHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+	sec, err := strconv.ParseInt(r.FormValue("seconds"), 10, 64)
+	if sec <= 0 || err != nil {
+		sec = 30
+	}
+
+	if durationExceedsWriteTimeout(r, float64(sec)) {
+		serveError(w, http.StatusBadRequest, "profile duration exceeds server's WriteTimeout")
+		return
+	}
+
+	// Set Content Type assuming StartCPUProfile will work,
+	// because if it does it starts writing.
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Disposition", `attachment; filename="profile"`)
+
+	pc := NewPprofAPICollector()
+	err = pc.StartCPUProfile(w)
+	if err != nil {
+		serveError(w, http.StatusInternalServerError, "Could not enable CPU profiling: "+err.Error())
+		return
+	}
+	// TODO: fix me.
+	// This can be fixed by always starts a 1 second profiling one by one,
+	// but to aggregate (merge) multiple profiles into one according to the precision.
+	//  |<-- 1s -->|
+	// -|----------|----------|----------|----------|----------|-----------|-----> Background profile task timeline.
+	//                            |________________________________|
+	//       (start cpu profile)  v                                v (stop cpu profile)    // expected profile timeline
+	//                        |________________________________|                           // actual profile timeline
+	time.Sleep(time.Second * time.Duration(sec))
+	err = pc.StopCPUProfile()
+	if err != nil {
+		serveError(w, http.StatusInternalServerError, "Could not enable CPU profiling: "+err.Error())
+	}
+}
+
+const (
+	labelSQLDigest  = "sql_digest"
+	labelPlanDigest = "plan_digest"
+)
+
+// PprofAPICollector is a cpu profile consumer for Pprof API usage.
+type PprofAPICollector struct {
+	ctx       context.Context
+	cancel    context.CancelFunc
+	started   atomic.Bool
+	firstRead chan struct{}
+	wg        sync.WaitGroup
+
+	dataCh   ProfileConsumer
+	profiles []*ProfileData
+	writer   io.Writer
+}
+
+// NewPprofAPICollector returns a new NewPprofAPICollector.
+func NewPprofAPICollector() *PprofAPICollector {
```

Comment:
This is a bad naming. From the perspective of `PprofAPICollector`'s user: what does it collect? Does it collect a API? Why do I need to collect a API?

How about just call it `NewCPUProfiler`? This API looks nice:

CPUProfiler::Start(w)
CPUProfiler::Stop()

Reply:
`CPUProfiler ` will conflict with `parallelCPUProfiler`? And it is just a consumer/collector, not really to do cpu profiling.
How about rename to `CPUProfileCollector`